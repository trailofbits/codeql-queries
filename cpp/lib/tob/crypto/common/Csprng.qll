import cpp


abstract class Csprng extends Function {
  abstract int getAStrongRandomnessSource();

  abstract int getRequestedBytes();
}


class CsprngCall extends FunctionCall {
  Csprng csprng;

  // Ensures that the call target is an instance of Csprng.
  CsprngCall() { csprng = this.getTarget() }

  Expr getAStrongRandomnessSource() { result = this.getArgument(csprng.getAStrongRandomnessSource()) }

  Expr getRequestedBytes() { result = this.getArgument(csprng.getRequestedBytes()) }
}
