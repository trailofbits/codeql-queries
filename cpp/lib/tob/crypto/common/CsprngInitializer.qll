import cpp
import StrongRandomnessSink


abstract class CsprngInitializer extends Function {
  // Returns the index of the seed argument (if it exists).
  abstract int getAStrongRandomnessSink();
}

class CsprngInitializerCall extends FunctionCall {
  CsprngInitializer init;
  
  CsprngInitializerCall() {
    this.getTarget() = init
  }

  // Returns the seed argument.
  Expr getAStrongRandomnessSink() {
    result = this.getArgument(init.getAStrongRandomnessSink())
  }
}

// A type representing the seed argument of a CSPRNG initializer function.
class CsprngInitializerSink extends StrongRandomnessSink {
  CsprngInitializerCall init;

  CsprngInitializerSink() {
    this = init.getAStrongRandomnessSink()
  }

  override string getDescription() {
    result = init.getTarget().getQualifiedName() + " parameter " + this.toString()
  }
}
