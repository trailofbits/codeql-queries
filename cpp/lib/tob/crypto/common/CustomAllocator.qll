import cpp
private import semmle.code.cpp.controlflow.StackVariableReachability


/**
 * A custom allocator which returns a pointer to the allocated object.
 */
abstract class CustomAllocator extends Function {
  CustomDeallocator dealloc;

  CustomDeallocator getDeallocator() {
    result = dealloc
  }

  // Override this and return the index of the pointer argument if the
  // allocated pointer is passed as an argument.
  int getPointer() {
    result = -1
  }
}

class CustomAllocatorCall extends FunctionCall {
  CustomAllocator alloc;

  CustomAllocatorCall() {
    this = alloc.getACallToThisFunction()
  }

  Expr getPointer() {
    if alloc.getPointer() < 0 then
      result = this
    else
      result = this.getArgument(alloc.getPointer())
  }

  CustomAllocator getAllocator() {
    result = alloc
  }

  CustomDeallocatorCall getADeallocatorCall() {
    result.getTarget() = alloc.getDeallocator()
  }
}

/**
 * A custom deallocator which takes the object to deallocate as an argument.
 */
abstract class CustomDeallocator extends Function {
  abstract int getPointer();
}

class CustomDeallocatorCall extends FunctionCall {
  CustomDeallocator dealloc;

  CustomDeallocatorCall() {
    this = dealloc.getACallToThisFunction()
  }

  Expr getPointer() {
    result = this.getArgument(dealloc.getPointer())
  }
}

class CustomAllocatorLeak extends StackVariableReachabilityWithReassignment {
  CustomAllocatorCall alloc;

  CustomAllocatorLeak() {
    this = "CustomAllocatorLeak"
  }

  override predicate isSourceActual(ControlFlowNode node, StackVariable var) {
    // A source is an allocation of the variable.
    node = alloc.getPointer() and (
      // When alloc returns allocated object.
      alloc.getPointer() = var.getAnAssignedValue() or
      // When alloc takes a pointer as an argument the variable can be either a
      // pointer or an object v, in which case &v is passed to the function and
      // we need to invoke getAChild() to get a variable access.
      alloc.getPointer().getAChild*() = var.getAnAccess()
    )
  }

  override predicate isBarrier(ControlFlowNode node, StackVariable var) {
    node = alloc.getADeallocatorCall() and
    node.(CustomDeallocatorCall).getPointer().getAChild*() = var.getAnAccess()
  }

  override predicate isSinkActual(ControlFlowNode node, StackVariable var) {
    // A sink is a return statement not returning the allocated variable.
    var.getFunction() = node.(ReturnStmt).getEnclosingFunction()
    and not node.(ReturnStmt).getExpr() = var.getAnAccess()
  }
}

class CustomAllocatorUseAfterFree extends StackVariableReachabilityWithReassignment {
  CustomAllocatorUseAfterFree() {
    this = "CustomAllocatorUseAfterFree"
  }

  override predicate isSourceActual(ControlFlowNode node, StackVariable var) {
    node.(CustomDeallocatorCall).getPointer() = var.getAnAccess()
  }

  override predicate isSinkActual(ControlFlowNode node, StackVariable var) {
    // A use is an access which is not a reassignment.
    node = var.getAnAccess() and not
    this.isAnAssignment(node, var)
  }

  override predicate isBarrier(ControlFlowNode node, StackVariable var) {
    // Stop tracking the variable if it is reassigned.
    this.isAnAssignment(node, var)
  }

  // Returns true if the node is the lvalue of an assignment of var.
  predicate isAnAssignment(ControlFlowNode node, StackVariable var) {
    node = var.getAnAssignment().getLValue()
  }
}
