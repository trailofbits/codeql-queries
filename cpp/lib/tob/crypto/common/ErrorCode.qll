import cpp

/**
 * A type modelling functions that return an error or status code that should be acted on.
 */
abstract class ReturnsErrorCode extends Function { }


/**
 * A type modelling the return value from a function that returns an error or status code.
 */
class ErrorCode extends FunctionCall {
  ErrorCode() { this.getTarget() instanceof ReturnsErrorCode }
}
