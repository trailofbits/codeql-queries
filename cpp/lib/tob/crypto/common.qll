import common.InitUpdateFinalPattern
import common.SymmetricCipherContext

import common.Csprng
import common.ErrorCode
import common.HashFunction
import common.CustomAllocator
import common.CsprngInitializer
import common.StrongPasswordSink
import common.StrongRandomnessSink
import common.WeakRandomnessSource
import common.StaticCryptoVariable
import common.KeyDerivationFunction
