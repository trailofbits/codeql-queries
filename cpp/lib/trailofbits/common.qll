import common.CustomAllocator
import common.EscapeAnalysis
import common.MethodCall
import common.ReturnValue