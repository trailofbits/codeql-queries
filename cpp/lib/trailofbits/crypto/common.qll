import trailofbits.crypto.common.InitUpdateFinalPattern
import trailofbits.crypto.common.SymmetricCipherContext

import trailofbits.crypto.common.Csprng
import trailofbits.crypto.common.ErrorCode
import trailofbits.crypto.common.HashFunction
import trailofbits.crypto.common.CustomAllocator
import trailofbits.crypto.common.CsprngInitializer
import trailofbits.crypto.common.StrongPasswordSink
import trailofbits.crypto.common.StrongRandomnessSink
import trailofbits.crypto.common.WeakRandomnessSource
import trailofbits.crypto.common.StaticCryptoVariable
import trailofbits.crypto.common.KeyDerivationFunction
