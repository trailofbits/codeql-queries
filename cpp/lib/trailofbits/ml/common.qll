import common.CustomAllocator
import common.ReturnValue