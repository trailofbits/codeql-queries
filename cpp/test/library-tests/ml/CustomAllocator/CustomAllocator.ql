import cpp
import trailofbits.ml.common
import trailofbits.ml.libraries

from 
    CustomAllocator allocator,
    StackVariable var,
    ControlFlowNode node,
    string message,
    int line
where 
    (
        allocator.isAllocatedBy(node, var) and
        message = "`" + var + "` is allocated here" and
        line = node.getLocation().getStartLine()
    ) or 
    (
        allocator.isFreedBy(node, var) and
        message = "`" + var + "` is freed here" and
        line = node.getLocation().getStartLine()
    )
select
    line, message