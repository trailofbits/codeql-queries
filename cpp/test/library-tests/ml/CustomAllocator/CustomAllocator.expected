| 8 | `graph` is allocated here |
| 16 | `graph` is freed here |
| 22 | `graph` is allocated here |
| 33 | `graph` is freed here |
| 35 | `graph` is freed here |
